src/coreclr/nativeaot/System.Private.CoreLib/src/System/ComAwareWeakReference.NativeAot.cs

```diff
@@ -2,32 +2,49 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System;
+using System.Runtime.CompilerServices;
 
 #if FEATURE_COMINTEROP || FEATURE_COMWRAPPERS
 
-#pragma warning disable IDE0060
-
 namespace System
 {
     internal sealed partial class ComAwareWeakReference
     {
-        internal static object? ComWeakRefToObject(IntPtr pComWeakRef, long wrapperId)
+        // We don't want to consult ComWrappers if no RCW objects have been created.
+        // In addition we don't want a direct reference to ComWrappers to allow for better
+        // trimming.  So we instead make use of delegates that ComWrappers registers when
+        // it is used.
+        private static unsafe volatile delegate*<IntPtr, long, object?> s_comWeakRefToObjectCallback;
+        private static unsafe volatile delegate*<object, bool> s_possiblyComObjectCallback;
+        private static unsafe volatile delegate*<object, out long, IntPtr> s_objectToComWeakRefCallback;
+
+        internal static unsafe void InitializeCallbacks(
+            delegate*<IntPtr, long, object?> comWeakRefToObject,
+            delegate*<object, bool> possiblyComObject,
+            delegate*<object, out long, IntPtr> objectToComWeakRef)
+        {
+            // PossiblyComObjectCallback is initialized last to avoid any potential race
+            // conditions where functions are being called while initialization is happening.
```

Comment:
I am looking at this again - what is the exact race condition that you are worried about? I do not see it. Could you please elaborate?

Reply:
The race condition I was concerned about is the callbacks being initialized while a weak reference was being allocated and _PossiblyComObjectCallback_ returning true when the other callbacks hadn't been set yet.  But thinking about it again, it shouldn't return true because due to where I initialize it, there won't be any RCWs yet.  So I will undo the volatile change and remove the comment.